pkg/controller/util/endpoint/controller_utils_test.go

```diff
@@ -224,3 +232,117 @@ func TestShouldPodBeInEndpoints(t *testing.T) {
 		}
 	}
 }
+
+func TestGetPodServiceMemberships(t *testing.T) {
+	cache := NewServiceSelectorCache()
+	pod0 := &v1.Pod{
+		Status: v1.PodStatus{
+			Phase: v1.PodRunning,
+			PodIP: "10.0.0.1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: "test",
+			Name:      "test-pod-0",
+			Labels: map[string]string{
+				"app":    "test-0",
+				"label0": "label-0",
+			},
+		},
+		Spec: v1.PodSpec{
+			NodeName: "node1",
+		},
+	}
+
+	fakeInformerFactory := informers.NewSharedInformerFactory(&fake.Clientset{}, 0*time.Second)
+	var serviceList []*v1.Service
+	for i := 0; i < 5; i++ {
+		service := &v1.Service{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:        fmt.Sprintf("service-%d", i),
+				Namespace:   "test",
+				Annotations: map[string]string{},
+			},
+			Spec: v1.ServiceSpec{
+				Selector: map[string]string{
+					"app": fmt.Sprintf("test-%d", i),
+				},
+			},
+			Status: v1.ServiceStatus{},
+		}
+		serviceList = append(serviceList, service)
+		fakeInformerFactory.Core().V1().Services().Informer().GetStore().Add(service)
+	}
+
+	expectServices := sets.String{}
+	expectServices.Insert("test/service-0")
+
+	services, err := cache.GetPodServiceMemberships(fakeInformerFactory.Core().V1().Services().Lister(), pod0)
+	if err != nil || !services.Equal(expectServices) {
+		t.Errorf("expected services: %v, got: %v, error: %v", expectServices, services, err)
+	}
+	for i := 0; i < 5; i++ {
+		key, err := controller.KeyFunc(serviceList[i])
+		if err != nil {
+			t.Errorf("get %v key with error %v", serviceList[i], err)
+		}
+		v, ok := cache.cache[key]
+		selector := v.(labels.Selector)
+		if !ok || !selector.Matches(labels.Set{"app": fmt.Sprintf("test-%d", i)}) {
+			t.Errorf("fail to cache service %s", key)
+		}
+	}
+
+	// test key which not exist in cache
+	if _, ok := cache.cache["test/service-5"]; ok {
+		t.Error("test_service-5 shouldn't exist in cache")
+	}
+}
+
+func TestServiceSelectorCache(t *testing.T) {
+	cache := NewServiceSelectorCache()
```

Comment:
This might also be easier to write with a list of test cases, potentially including the following values:

- setKey
- setSelector (map[string]string)
- deleteKey
- getKey
- expectedSelector

This would allow each test case to go through optional set, delete, get flows or similar to show the functionality of each cache function. That's of course dependent on changing the `Update` function above to use `service.Spec.Selector` as the argument. 

I'm not 100% sold on this specific structure for the test, but I do think something that allowed more, simpler, test cases to cover this function would be helpful.

Reply:
All addressed, appreciate for your careful review.